mp4parse_capi/src/lib.rs

```diff
@@ -144,36 +149,134 @@ impl Default for Mp4ParseEncryptionSchemeType {
     }
 }
 
+/// A zero-overhead wrapper around integer types for the sake of always
+/// requiring checked arithmetic
+#[repr(transparent)]
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct CheckedInteger<T>(pub T);
+
+impl<T> From<T> for CheckedInteger<T> {
+    fn from(i: T) -> Self {
+        Self(i)
+    }
+}
+
+// Orphan rules prevent a more general implementation, but this suffices
+impl From<CheckedInteger<i64>> for i64 {
+    fn from(checked: CheckedInteger<i64>) -> i64 {
+        checked.0
+    }
+}
+
+impl<T, U: Into<T>> Add<U> for CheckedInteger<T>
+where
+    T: CheckedAdd,
+{
+    type Output = Option<Self>;
+
+    fn add(self, other: U) -> Self::Output {
+        self.0.checked_add(&other.into()).map(Into::into)
+    }
+}
+
+impl<T, U: Into<T>> Sub<U> for CheckedInteger<T>
+where
+    T: CheckedSub,
+{
+    type Output = Option<Self>;
+
+    fn sub(self, other: U) -> Self::Output {
+        self.0.checked_sub(&other.into()).map(Into::into)
+    }
+}
+
+/// Implement subtraction of checked `u64`s returning i64
+// This is necessary for handling Mp4parseTrackInfo::media_time gracefully
+impl Sub for CheckedInteger<u64> {
+    type Output = Option<CheckedInteger<i64>>;
+
+    fn sub(self, other: Self) -> Self::Output {
+        if self >= other {
+            self.0
+                .checked_sub(other.0)
+                .and_then(|u| i64::try_from(u).ok())
+                .map(CheckedInteger)
+        } else {
+            other
+                .0
+                .checked_sub(self.0)
+                .and_then(|u| i64::try_from(u).ok())
+                .map(i64::neg)
+                .map(CheckedInteger)
+        }
```

Comment:
It occurs to me that this could be expressed a little less redundantly as:
```rust
        let maybe_negate: fn(i64) -> CheckedInteger<i64>;
        if self >= other {
            maybe_negate = CheckedInteger;
            self.0.checked_sub(other.0)
        } else {
            maybe_negate = |d| CheckedInteger(-d);
            other.0.checked_sub(self.0)
        }
        .and_then(|u| i64::try_from(u).ok())
        .map(maybe_negate)
```

Do y'all find that to be better or worse in terms of readability?